pkg/agent/openflow/network_policy.go

```diff
@@ -706,53 +709,63 @@ func (c *client) NewDNSPacketInConjunction(id uint32) error {
 	if err := c.ofEntryOperations.AddAll(conj.actionFlows); err != nil {
 		return fmt.Errorf("error when adding action flows for the DNS conjunction: %w", err)
 	}
-	udpService := v1beta2.Service{
-		Protocol: &protocolUDP,
-		SrcPort:  &dnsPort,
-	}
 	dnsPriority := priorityDNSIntercept
 	conj.serviceClause = conj.newClause(1, 2, getTableByID(conj.ruleTableID), nil)
 	conj.toClause = conj.newClause(2, 2, getTableByID(conj.ruleTableID), nil)
 	c.featureNetworkPolicy.conjMatchFlowLock.Lock()
 	defer c.featureNetworkPolicy.conjMatchFlowLock.Unlock()
-	ctxChanges := conj.serviceClause.addServiceFlows(c.featureNetworkPolicy, []v1beta2.Service{udpService}, &dnsPriority, false)
-
-	tcpFlags := TCPFlags{
-		// URG|ACK|PSH|RST|SYN|FIN|
-		Flag: 0b011000,
-		Mask: 0b011000,
-	}
-	tcpDNSPort := types.BitRange{Value: uint16(dnsPort)}
+	var ctxChanges []*conjMatchFlowContextChange
 	for _, proto := range c.featureNetworkPolicy.ipProtocols {
-		tcpServiceMatch := &conjunctiveMatch{
+		tcpMatch := &conjunctiveMatch{
+			tableID:  conj.serviceClause.ruleTable.GetID(),
+			priority: &dnsPriority,
+		}
+		udpMatch := &conjunctiveMatch{
 			tableID:  conj.serviceClause.ruleTable.GetID(),
 			priority: &dnsPriority,
 		}
 		if proto == binding.ProtocolIP {
-			tcpServiceMatch.matchPairs = []matchPair{
+			tcpMatch.matchPairs = []matchPair{
 				{
 					matchKey:   MatchTCPSrcPort,
-					matchValue: tcpDNSPort,
+					matchValue: types.BitRange{Value: uint16(dnsPort)},
 				},
 				{
-					matchKey:   MatchTCPFlags,
-					matchValue: tcpFlags,
+					matchKey: MatchCTStateRpl,
+				},
+			}
+			udpMatch.matchPairs = []matchPair{
+				{
+					matchKey:   MatchUDPSrcPort,
+					matchValue: types.BitRange{Value: uint16(dnsPort)},
+				},
+				{
+					matchKey: MatchCTStateRpl,
 				},
 			}
 		} else if proto == binding.ProtocolIPv6 {
-			tcpServiceMatch.matchPairs = []matchPair{
+			tcpMatch.matchPairs = []matchPair{
 				{
 					matchKey:   MatchTCPv6SrcPort,
-					matchValue: tcpDNSPort,
+					matchValue: types.BitRange{Value: uint16(dnsPort)},
+				},
+				{
+					matchKey: MatchCTStateRpl,
+				},
+			}
+			udpMatch.matchPairs = []matchPair{
+				{
+					matchKey:   MatchUDPv6SrcPort,
+					matchValue: types.BitRange{Value: uint16(dnsPort)},
 				},
 				{
-					matchKey:   MatchTCPv6Flags,
-					matchValue: tcpFlags,
+					matchKey: MatchCTStateRpl,
```

Comment:
Why is this condition needed in UDP ?

Reply:
We want to make sure we only send solicited DNS responses to userspace.

Reply:
Better to add comments for the information?

Reply:
Added.